vmware_dvs/utils/dvs_util.py

```diff
@@ -457,89 +454,120 @@ def _valid_uuid(self, name):
+                self._pg_cache[pg_name]['status'] != READY_PG_STATUS):
+            sleep(TASK_POOL_INTERVAL)
+
+    def create_network(self, network, segment):
+        name = self._get_net_name(network)
+        if self._pg_cache.get(name, {}).get('status') == CREATING_PG_STATUS:
+            self._wait_concurrent_pg_operation(name)
+            return self._pg_cache[name]['item']
+
+        prev_pg_status = self._pg_cache.get(name, {}).get('status')
+        self._pg_cache.setdefault(name, {}).update(
+            {'status': CREATING_PG_STATUS})
+        try:
+            pg = super(DVSControllerWithCache, self).create_network(
+                network, segment)
+        except exceptions.VMWareDVSException as e:
+            self._pg_cache[name].update({'status': prev_pg_status})
+            raise e
         else:
-            spec.inherited = '1'
-            spec.value = 'false'
-        return spec
+            self._pg_cache[name].update({
+                'item': pg,
+                'status': READY_PG_STATUS,
+                'pg_key': pg.value
+            })
+            return pg
+
+    def delete_network(self, network):
+        name = self._get_net_name(network)
+        try:
+            pg_ref = self._get_pg_by_name(name)
+        except exceptions.PortGroupNotFound:
+            self._pg_cache.pop(name, None)
+            return
+        self._delete_port_group(pg_ref, name)
+
+    def _delete_port_group(self, pg_ref, name):
+        # TODO(ekosareva): what should we do when pg is not in cache???
+        self._pg_cache[name].update({'status': REMOVING_PG_STATUS})
+        super(DVSControllerWithCache, self)._delete_port_group(pg_ref, name)
+        self._pg_cache.pop(name, None)
+
+    def _get_pg_by_name(self, pg_name):
+        if self._pg_cache.get(pg_name, {}).get('status') in (
+            READY_PG_STATUS, UPDATING_PG_STATUS, REMOVING_PG_STATUS):
+                return self._pg_cache[pg_name]['item']
+        raise exceptions.PortGroupNotFound(pg_name=pg_name)
+
+    def _increase_ports_on_portgroup(self, port_group):
+        pg_name = next((name for name, pg in self._pg_cache.iteritems()
+                        if pg['pg_key'] == port_group.value), None)
+        if (pg_name and
+                self._pg_cache[pg_name].get('status') == UPDATING_PG_STATUS):
+            self._wait_concurrent_pg_operation(pg_name)
+            return
+
+        self._pg_cache.setdefault(pg_name, {}).update(
+            {'status': UPDATING_PG_STATUS})
+        try:
+            super(DVSControllerWithCache, self).\
+                _increase_ports_on_portgroup(port_group)
+        finally:
+            self._pg_cache[pg_name].update({'status': READY_PG_STATUS})
+
+    def _lookup_unbound_port(self, port_group):
+        pg_name = next((name for name, pg in self._pg_cache.iteritems()
+                        if pg['pg_key'] == port_group.value), None)
+
+        #TODO(ekosareva): if we didn't find pg_name should we raise exception?
```

Comment:
It will be the bug in cache functionality. It needs to return port key in any case if it is possible. Agent can create PG if it does not exist or find if does